metrics/runtime_metrics.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// RuntimeConfig contains configuration for initializing runtime metrics
+type RuntimeConfig struct {
+	Disabled bool `yaml:"disabled"`
+}
+
+// NewRuntimeCollector creates a new RuntimeCollector.
+func NewRuntimeCollector(
+	scope tally.Scope, collectInterval time.Duration,
+) *RuntimeCollector {
+	var memstats runtime.MemStats
+	runtime.ReadMemStats(&memstats)
+	return &RuntimeCollector{
+		scope:           scope,
+		collectInterval: collectInterval,
+		metrics: runtimeMetrics{
+			numGoRoutines:   scope.Gauge("num-goroutines"),
+			goMaxProcs:      scope.Gauge("gomaxprocs"),
+			memoryAllocated: scope.Gauge("memory.allocated"),
+			memoryHeap:      scope.Gauge("memory.heap"),
+			memoryHeapIdle:  scope.Gauge("memory.heapidle"),
+			memoryHeapInuse: scope.Gauge("memory.heapinuse"),
+			memoryStack:     scope.Gauge("memory.stack"),
+			numGC:           scope.Counter("memory.num-gc"),
+			gcPauseMs:       scope.Timer("memory.gc-pause-ms"),
+			lastNumGC:       memstats.NumGC,
+		},
+		started: false,
+		quit:    make(chan struct{}),
+	}
+}
+
+// IsRunning returns true if the collector has been started and false if not.
+func (r *RuntimeCollector) IsRunning() bool {
+	r.startedMutex.RLock()
+	defer r.startedMutex.RUnlock()
+	return r.started
+}
+
+// Start starts the collector thread that periodically emits metrics.
+func (r *RuntimeCollector) Start() {
+	r.startedMutex.RLock()
+	if r.started {
+		r.startedMutex.RUnlock()
+		return
+	}
+	r.startedMutex.RUnlock()
+	go func() {
+		ticker := time.NewTicker(r.collectInterval)
+		for {
+			select {
+			case <-ticker.C:
+				r.generate()
+			case <-r.quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	r.startedMutex.Lock()
+	r.started = true
+	r.startedMutex.Unlock()
+}
+
+// generate sends runtime metrics to the local metrics collector.
+func (r *RuntimeCollector) generate() {
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+
+	r.metrics.numGoRoutines.Update(int64(runtime.NumGoroutine()))
+	r.metrics.goMaxProcs.Update(int64(runtime.GOMAXPROCS(0)))
+	r.metrics.memoryAllocated.Update(int64(memStats.Alloc))
+	r.metrics.memoryHeap.Update(int64(memStats.HeapAlloc))
+	r.metrics.memoryHeapIdle.Update(int64(memStats.HeapIdle))
+	r.metrics.memoryHeapInuse.Update(int64(memStats.HeapInuse))
+	r.metrics.memoryStack.Update(int64(memStats.StackInuse))
+
+	// memStats.NumGC is a perpetually incrementing counter (unless it wraps at 2^32)
+	num := memStats.NumGC
+	lastNum := atomic.SwapUint32(&r.metrics.lastNumGC, num) // reset for the next iteration
+	if delta := num - lastNum; delta > 0 {
+		r.metrics.numGC.Inc(int64(delta))
+		// Match the MemStats buffer and generate only the last _numGCThreshold
+		if delta >= _numGCThreshold {
+			lastNum = num - _numGCThreshold
+		}
+		for i := lastNum; i != num; i++ {
+			pause := memStats.PauseNs[i%256]
+			r.metrics.gcPauseMs.Record(time.Duration(pause))
+		}
+	}
+}
+
+// Close stops collecting runtime metrics. It cannot be started again after it's
```

Comment:
@anuptalwalkar maybe we should consider this strategy for service start/stop... might make handling states and cleanup a lot easier. If a service is stopped - that's it. Re-create it